third_party/pyth/xc-governance-sdk-js/src/instructions.ts

```diff
@@ -0,0 +1,174 @@
+import { ChainId } from "@certusone/wormhole-sdk";
+
+import { SerializeUtils, Serializable } from "./serialize";
+
+enum Module {
+  Executor = 0,
+  Target,
+}
+
+enum TargetAction {
+  UpgradeContract = 0,
+  SetGovernanceDataSource,
+  SetDataSources,
+  SetFee,
+  SetValidPeriod,
+}
+
+abstract class HexString implements Serializable {
+  private readonly addressBuffer: Buffer;
+
+  constructor(address: string, byteLen: number) {
+    if (address.startsWith("0x")) {
+      address = address.substring(2);
+    }
+    if (address.length !== 2 * byteLen) {
+      throw new Error(`Expected address of length ${2 * byteLen}, found ${address.length}`);
+    }
+    this.addressBuffer = Buffer.from(address, "hex");
+    if (this.addressBuffer.length === 0) {
+      throw new Error(`Given address is not in hex format`);
+    }
+  }
+
+  serialize(): Uint8Array {
+    return Uint8Array.from(this.addressBuffer.values());
+  }
+}
+
+export class HexString20Bytes extends HexString {
+  constructor(address: string) {
+    super(address, 20);
+  }
+}
+
+export class HexString32Bytes extends HexString {
+  constructor(address: string) {
+    super(address, 32);
+  }
+}
+
+export class DataSource implements Serializable {
+  constructor(
+    private readonly emitterChain: ChainId,
+    private readonly emitterAddress: HexString32Bytes,
+  ) { };
+
+  serialize(): Uint8Array {
+    return SerializeUtils.concat(
+      SerializeUtils.serializeUint16(Number(this.emitterChain)),
+      this.emitterAddress.serialize()
+    );
+  }
+}
+
+// Magic is `PTGM` encoded as a 4 byte data: Pyth Governance Message
+const MAGIC: number = 0x5054474d;
+
+abstract class Instruction implements Serializable {
+  constructor(
+    private module: Module,
+    private action: number,
+    private targetChainId: ChainId,
+  ) { };
+
+  protected abstract serializePayload(): Uint8Array;
+
+  private serializeHeader(): Uint8Array {
+    return SerializeUtils.concat(
+      SerializeUtils.serializeUint32(MAGIC),
```

Comment:
is there really not an existing library to serialize binary messages like this?

(I know we're in a rush, so don't worry about this for this PR. It's not a big deal right now.)

Reply:
also fwiw, for this kind of use case, usually the right pattern is to make some kind of `Builder` object that stores a mutable buffer and has methods to add stuff to the end. Like:

```
let builder =BufferBuilder()
builder.addUint32(...)
builder.addUint64(...)
return builder.build()
```

again though, don't worry about this right now. this is purely an FYI.

Reply:
There are some utils in `ethers` that worm CLI uses. But I didn't like their approach and tried to stick to the native ways. In a quick search I couldn't find cleaner ways. 

I couldn't find a mutable way to do this and created concat to be more efficient (copy only once). But it's not so important. Initially I thought builder here could be a bit annoying but now it seems good to me. I'll change it.

Reply:
I figured out Buffer has good write methods and I used them and created BufferBuilder. Now it's much clear. I tried to use `mutable-buffer` package but it didn't have 64bit writes. Other packages were too complex for this use case.